service/src/main/java/com/ericsson/eiffel/remrem/generate/controller/RemremGenerateController.java

```diff
@@ -99,64 +98,158 @@ public void setRestTemplate(RestTemplate restTemplate) {
+        JsonObject errorResponse = new JsonObject();
+        String exceptionMessage = e.getMessage();
+        if (e instanceof REMGenerateException) {
+            List<HttpStatus> statuses = List.of(
+                    HttpStatus.NOT_ACCEPTABLE, HttpStatus.EXPECTATION_FAILED, HttpStatus.SERVICE_UNAVAILABLE, HttpStatus.UNPROCESSABLE_ENTITY
+            );
+            for (HttpStatus status : statuses) {
+                if (exceptionMessage.contains(Integer.toString(status.value()))) {
+                    return createResponseEntity(
+                            status, e.getMessage(), "fail", errorResponse);
+                }
             }
-            else {
-                return new ResponseEntity<>(parser.parse(e1.getMessage()), HttpStatus.UNPROCESSABLE_ENTITY);
+            return createResponseEntity(HttpStatus.BAD_REQUEST, e.getMessage(), "fail", errorResponse);
+        } else if (e instanceof JsonSyntaxException) {
+            log.error("Failed to parse JSON: ", exceptionMessage);
+            return createResponseEntity(HttpStatus.BAD_REQUEST, e.getMessage(), "fail", errorResponse);
+        } else {
+            log.error("Unexpected exception caught", exceptionMessage);
+            return createResponseEntity(HttpStatus.INTERNAL_SERVER_ERROR, exceptionMessage, "fail", errorResponse);
+        }
+    }
+
+    /**
+     * This helper method basically generate or process one event
+     * @return JsonObject generated event
+     */
+
+    public JsonObject processEvent(String msgProtocol, String msgType, Boolean failIfMultipleFound,
+                                   Boolean failIfNoneFound, Boolean lookupInExternalERs, int lookupLimit,
+                                   Boolean okToLeaveOutInvalidOptionalFields, JsonObject jsonObject) throws REMGenerateException, JsonSyntaxException {
+        JsonObject eventResponse = new JsonObject();
+        JsonElement parsedResponse;
+
+        JsonObject event = erLookup(jsonObject, failIfMultipleFound, failIfNoneFound, lookupInExternalERs, lookupLimit);
+        MsgService msgService = getMessageService(msgProtocol);
+
+        if (msgService != null) {
+            String response = msgService.generateMsg(msgType, event, isLenientEnabled(okToLeaveOutInvalidOptionalFields));
+            parsedResponse = JsonParser.parseString(response);
+            JsonObject parsedJson = parsedResponse.getAsJsonObject();
+
+            if (parsedJson.has(JSON_ERROR_MESSAGE_FIELD)) {
+                createResponseEntity(HttpStatus.BAD_REQUEST, "fail", TEMPLATE_ERROR, eventResponse);
+                return eventResponse;
+            } else {
+                return parsedJson;
             }
-        } catch (Exception e) {
-            log.error("Unexpected exception caught", e);
-            return new ResponseEntity<>(parser.parse(RemremGenerateServiceConstants.INTERNAL_SERVER_ERROR),
-                    HttpStatus.INTERNAL_SERVER_ERROR);
         }
+        return eventResponse;
```

Comment:
We cannot return an empty JSON object. Situation when msgService == null must be handled.